cloudbrain/rt_server/rt_server.py

```diff
@@ -0,0 +1,203 @@
+# -*- coding: utf-8 -*-
+
+from cloudbrain.settings import RABBITMQ_ADDRESS
+
+import pika
+import json
+import logging
+
+from sockjs.tornado.conn import SockJSConnection
+from sockjs.tornado import SockJSRouter
+from tornado.ioloop import PeriodicCallback, IOLoop
+from tornado.web import RequestHandler, Application
+
+from cloudbrain.subscribers.SubscriberInterface import Subscriber
+
+SERVER_PORT = 31415
+logging.getLogger().setLevel(logging.INFO)
+
+
+
+class RtStreamConnection(SockJSConnection):
+    """RtStreamConnection connection implementation"""
+    # Class level variable
+    clients = set()
+
+    def __init__(self, session):
+        super(self.__class__, self).__init__(session)
+        self.subscribers = {}
+
+
+    def send_probe_factory(self, metric_name):
+
+        def send_probe(body):
+            logging.debug("GOT: " + body)
+            buffer_content = json.loads(body)
+
+            for record in buffer_content:
+                record["metric"] = metric_name
+                self.send(json.dumps(record))
+
+        return send_probe
+
+
+    def on_open(self, info):
+        logging.info("Got a new connection...")
+        self.clients.add(self)
+        # self.timeout = PeriodicCallback(self.send_heartbeat, 1000)
+        # self.timeout.start()
+
+    # This will receive instructions from the client to change the
+    # stream. After the connection is established we expect to receive a JSON
+    # with deviceName, deviceId, metric; then we subscribe to RabbitMQ and
+    # start streaming the data.
+    #
+    # NOTE: it's not possible to open multiple connections from the same client.
+    #       so in case we need to stream different devices/metrics/etc. at the
+    #       same time, we need to use a solution that is like the multiplexing
+    #       in the sockjs-tornado examples folder.
+    def on_message(self, message):
+        logging.info("Got a new message: " + message)
+
+        stream_configuration = json.loads(message)
+        device_name = stream_configuration['deviceName']
+        device_id = stream_configuration['deviceId']
+        metric = stream_configuration['metric']
+
+        if metric not in self.subscribers:
+            self.subscribers[metric] = TornadoSubscriber(callback=self.send_probe_factory(metric),
+                                       device_name=device_name,
+                                       device_id=device_id,
+                                       rabbitmq_address=RABBITMQ_ADDRESS,
+                                       metric_name=metric)
+
+
+            self.subscribers[metric].connect()
+
+
+
+    def on_close(self):
+        if self.subscriber is not None:
```

Comment:
This is the `on_close` event of the web-socket connection itself, so we should probably loop and close all the `subscribers` since they can be more than one now. Moreover, we need to pay attention catching exception in case the close fails, to avoid to just close some of them and keeping other open but lingering around.